feat(authz): Authorino for Service Mesh

israel-hdez opened the pull request:
## Description

This first iteration is to cover authentication needs for KServe

* Add templates to install Authorino
* Add templates to configure Service Mesh to use Authorino to delegate Authorization
* Add KServe-specific templates add ability to secure KServe Inference Services
* Add relevant fields to DSCInitialization resource
* Code for proper cleanup, in case of uninstalling

Most (if not all) of this code comes from pull request opendatahub-io/opendatahub-operator#605. Attribution to original authors: @bartoszmajsak, @aslakknutsen, @cam-garrison, et. al.

Related opendatahub-io/kserve#128
Fixes https://issues.redhat.com/browse/RHOAIENG-1856

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

TODO

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

zdtsw commented:
are we gonna have 605 for UI+workbench and this for kserve?

bartoszmajsak commented on infrastructure/v1/servicemesh_types.go:
```diff
@@ -38,3 +41,30 @@ type IngressGatewaySpec struct {
+	// Kubernetes apiserver (kubernetes.default.svc).
+	// +kubebuilder:default={"https://kubernetes.default.svc"}
+	Audiences []string `json:"audiences,omitempty"`
+	// Label narrows amount of AuthConfigs to process by Authorino service.
+	// +kubebuilder:default=authorino/topic=odh
+	Label string `json:"label,omitempty"`
+	// Image allows to define a custom container image to be used when deploying Authorino's instance.
+	// +kubebuilder:default="quay.io/kuadrant/authorino:v0.16.0"
+	Image string `json:"image,omitempty"`
```
These are internal things, based on initial feedback from #605 we should not expose it directly in this API

israel-hdez replied:
I was going to add and ENV flag to disable by default this.

What I've learnt in the past is that these settings are useful not only for dev, but also for people who deploy a private registry where images are mirrored.

So, I agree the operator should provide non-customizable image on a standard install. Yet, it is still quite useful to provide a way to set your own images (not only for development). But the intent must be explicit.

israel-hdez commented:
@zdtsw 

> are we gonna have 605 for UI+workbench and this for kserve? for the overlapping part, should we have 605 in first then rebase for this one?

The plan was to merge this first, and rebase 605. However, we are reviewing the direction. Hold-on for now.

israel-hdez commented on pkg/deploy/setup.go:
```diff
@@ -89,3 +92,14 @@ func GetPlatform(cli client.Client) (Platform, error) {
 }
+
+// ShouldConfigureServiceMesh determines if the operator should invoke service-mesh specific setup.
+func ShouldConfigureServiceMesh(cli client.Client, dscispec *dsci.DSCInitializationSpec) (bool, error) {
```
I'm still a little bit dubious about this func.
Since we will be downstreaming this feature, I'm not sure what would be the platforms this should work for.

israel-hdez replied:
Removed

israel-hdez commented on controllers/dscinitialization/servicemesh_setup.go:
```diff
@@ -118,5 +118,48 @@ func configureServiceMeshFeatures(s *feature.FeaturesInitializer) error {
+		Manifests(
+			path.Join(rootDir, feature.AuthDir, "auth-smm.tmpl"),
+			path.Join(rootDir, feature.AuthDir, "base"),
+			// path.Join(rootDir, feature.AuthDir, "rbac"),
```
@bartoszmajsak Looks like we don't need this for the time being?

bartoszmajsak replied:
After looking at it I am not quite sure where this came from :) I assume SAR calls are working fine on your end, and generally speaking those roles should be granted by the operator itself, so I would say let's ditch it.

israel-hdez replied:
Removed.

israel-hdez commented on pkg/feature/builder.go:
```diff
@@ -32,6 +32,11 @@ func (fb *featureBuilder) For(spec *v1.DSCInitializationSpec) *featureBuilder {
 			AppNamespace:    spec.ApplicationsNamespace,
+			AuthorinoConfigs: AuthorinoConfigs{
+				CrName:         "authorino-mesh-authz-provider",
+				ContainerImage: "quay.io/kuadrant/authorino:v0.16.0",
```
Isn't it better to let the authorino operator use whatever is the default image?

bartoszmajsak replied:
Agree, let's remove it. It shouldn't be hardcoded like this. We will need to find a clean way of changing it, but this can be a part of a parallel work regarding installation of the Authorino controller itself. /cc @cam-garrison

bartoszmajsak commented on pkg/feature/feature.go:
```diff
@@ -161,6 +161,15 @@ func (f *Feature) addCleanup(cleanupFuncs ...Action) {
 	f.cleanups = append(f.cleanups, cleanupFuncs...)
 }
 
+func (f *Feature) ApplyManifest(filename string) error {
+	m := loadManifestFrom(filename)
+	if err := m.processTemplate(f.Spec); err != nil {
+		return err
+	}
+
+	return f.apply(m)
+}
```
With #797 coming in, especially considering that `fsys` is part of `manifest` struct now, we will need to have a look at how to adjust it @cam-garrison

bartoszmajsak commented on pkg/feature/servicemesh/cleanup.go:
```diff
@@ -0,0 +1,60 @@
+		return err
+	}
+	if !found {
+		log.Info("no extension providers found", "f", f.Name, "control-plane", mesh.Name, "namespace", mesh.Namespace)
```
With #805 log becomes part of `Feature` struct. Depending on which PR comes first we will have to adjust it. It will make `"f", f.Name` redundant. And on that note, for now:

```suggestion
		log.Info("no extension providers found", "feature", f.Name, "control-plane", mesh.Name, "namespace", mesh.Namespace)
```

israel-hdez replied:
Fixed

bartoszmajsak commented on pkg/feature/templates/servicemesh/authorino/base/operator-cluster-wide-no-tls.tmpl:
```diff
@@ -0,0 +1,15 @@
+  name: {{ .AuthorinoConfigs.CrName }}
+  namespace: {{ .Auth.Namespace }}
+spec:
+  image: {{ .AuthorinoConfigs.ContainerImage }}
```
As previously suggested - let's remove it for now.

israel-hdez replied:
Removed

bartoszmajsak commented on pkg/feature/types.go:
```diff
@@ -27,6 +28,12 @@ type OAuth struct {
+type AuthorinoConfigs struct {
+	CrName,
+	ContainerImage,
+	LabelSelectors string
```
If it's used only in one place - lets hardcode it in the template instead? I am not sure we need to make it configurable at this point. That was too much of a speculative development on my end.

israel-hdez replied:
OK, I'll hardcode it in the template.

israel-hdez replied:
Done

bartoszmajsak commented on pkg/feature/types.go:
```diff
@@ -27,6 +28,12 @@ type OAuth struct {
 }
 
+type AuthorinoConfigs struct {
+	CrName,
```
How about simply calling it `Name`?

israel-hdez replied:
`InstanceName`?
Just joking... Let's call it `Name`.

israel-hdez replied:
Renamed...

bartoszmajsak reviewed (changes requested):
Thanks for skillfully refining #605 and contributing those parts to incubation, @israel-hdez! 😊

I've reflected on the (mostly mine) code from a broader time perspective and shared a few insights. It would be fantastic if you could help tidy up the areas where I've left a bit of a mess behind.

bartoszmajsak commented:
Added @aslakknutsen as a reviewer so he can double check if service mesh resources for kserve are in the right state.

bartoszmajsak commented:
@zdtsw this PR depends on https://github.com/opendatahub-io/odh-model-controller/pull/130

israel-hdez commented on infrastructure/v1/servicemesh_types.go:
```diff
@@ -38,3 +41,21 @@ type IngressGatewaySpec struct {
+type AuthSpec struct {
+	// Namespace where it is deployed.
+	// +kubebuilder:default=auth-provider
+	Namespace string `json:"namespace,omitempty"`
```
I was wondering if it needs to be on its own namespace, or if we can re-use the `AppsNamespace`?

bartoszmajsak replied:
I would probably stick to keeping it separate considering this being an infra component like istio? But I don't have strict preference here, so I hope others can chime in :)

cam-garrison replied:
@guicassolato any thoughts on this? If we should be deploying authorino in its own separate namespace.

guicassolato replied:
While missing some of the details I'm sure, the decision of deploying in its own namespace or within the same namespace as the services it will help protect and where I imagine the AuthConfig CRs will be created, is not arbitrary. It depends first on the [reconciliation mode](https://docs.kuadrant.io/authorino/docs/architecture/#cluster-wide-vs-namespaced-instances) of the instance, an on the NetworkPolicies you want to set as well.

I see you are [using **cluster-wide** mode](https://github.com/opendatahub-io/opendatahub-operator/blob/3043c646ad9c41aeca22e6c1262d4d4b07961827/pkg/feature/templates/servicemesh/authorino/base/operator-cluster-wide-no-tls.tmpl#L9). In this case, the Authorino namespace does not matter for the purpose of visibility over the AuthConfigs. If you plan to have more than one instance or Authorino per cluster, including possibly the one from #605 (not sure if it's the same cluster), then you will want to assign the AuthConfigs specifically for the instance that is wired in the proxies to be activated for those hosts, which I see [you are doing](https://github.com/opendatahub-io/opendatahub-operator/blob/3043c646ad9c41aeca22e6c1262d4d4b07961827/pkg/feature/templates/servicemesh/authorino/base/operator-cluster-wide-no-tls.tmpl#L8).

israel-hdez replied:
Now, I'd lean towards installing to `AppsNamespace` assuming this is going to be *our* instance, and non-customizable by users.

bartoszmajsak replied:
I forgot to comment about it yesterday... I agree, let's keep it in the same place as the other components/controllers. This simplifies things.

Thanks @guicassolato for chiming in!

israel-hdez replied:
@bartoszmajsak So, I removed this field and used `AppNamespace` in the templates... Not sure if this is fair. But, now, we just have a struct inside a struct...

If you think this is not OK, I can restore the field.

israel-hdez commented on components/kserve/kserve.go:
```diff
@@ -229,3 +233,53 @@ func checkDepedentOps(cli client.Client) *multierror.Error {
 	return multiErr
 }
+
+func (k *Kserve) configureServiceMesh(cli client.Client, dscispec *dsciv1.DSCInitializationSpec, enabled bool) error {
```
I think both suggestions are mutually exclusive, since by moving the code to `servicemesh_setup.go` would require the `enabled` argument (since it is two different `packages`).

Moving the code can be debatable, since mesh setup for KServe, while the other file is mesh setup barebones?

For now, I just applied the first suggestion. Not sure about the 2nd one.

bartoszmajsak replied:
@israel-hdez what I meant is to move this part to its own file in this package instead of keeping it in in `kserve.go`.

israel-hdez replied:
Ah, that makes sense! I followed your suggestion.

israel-hdez commented on pkg/feature/servicemesh/cleanup.go:
```diff
@@ -0,0 +1,60 @@
+	for i, v := range extensionProviders {
+		extensionProvider, ok := v.(map[string]interface{})
+		if !ok {
+			fmt.Println("Unexpected type for extensionProvider")
```
Fixed... Although, check if you are happy with it.

bartoszmajsak replied:
I'm the most positive Pole you have ever met! :D

![image](https://github.com/opendatahub-io/opendatahub-operator/assets/719616/e0723133-beef-46f9-96ef-3acac03deb6e)

I like that you have thought about `WARN`. Let's figure out how to make it more explicit in the code.

israel-hdez commented:
@bartoszmajsak Ready for another review round.

israel-hdez commented on pkg/feature/templates/servicemesh/kserve/kserve-predictor-authorizationpolicy.tmpl:
```diff
@@ -0,0 +1,18 @@
+    - operation:
+        notPaths:
+        - /healthz*
+        - /metrics
```
@Xaenalt @VedantMahabaleshwarkar With the pod behind auth, we are leaving unprotected only these two paths/patterns.

I wonder if these are the only two that would need to remain open, or if there is anything else?

ReToCode replied:
I checked what we have on Queue-Proxy:

- 8022: `/wait-for-drain` (for a K8s hook)
- 9090: `/metrics`
- 8008: `/debug/pprof/` (if enabled)

Health checks (readiness-checks) are based on a header, so `/healthz*` is not going to match. But probably that is handled (rewritten by istio)?

```yaml
      readinessProbe:
        failureThreshold: 3
        httpGet:
          httpHeaders:
            - name: K-Network-Probe
              value: queue
          path: /
          port: 8012
          scheme: HTTP
        periodSeconds: 10
        successThreshold: 1
        timeoutSeconds: 1
```

israel-hdez replied:
So, I adjusted this Policy to leave unprotected these three paths.

I'm a little bit unsure about `/metrics`. I'm not sure if that would be considered as sensitive data... Not sure if metrics and their labels would ever contain something.

@bartoszmajsak @aslakknutsen Not sure if we should create a secondary `DENY` policy that would reject access to `/metrics` if called from the outside.

BTW, I removed `/healthz*`. Looks like this one is for the istio sidecar, but seems to work without it being here.

> Health checks (readiness-checks) are based on a header, so /healthz* is not going to match. But probably that is handled (rewritten by istio)

@ReToCode Yes. And looks like istio lets these requests go in without them being here in this AuthorizationPolicy.

ReToCode replied:
`/metrics` is called from Knative as well. Do our control-plane components send auth contexts? Otherwise this should also be unprotected (same for prometheus scraping).

israel-hdez replied:
> Do our control-plane components send auth contexts?

I think Nope. And yes, it's being left unprotected for now.

skonto replied:
For probes Istio rewrites them if you use for example: 
`sidecar.istio.io/rewriteAppHTTPProbers`, I think you use the latter correct?

israel-hdez replied:
I saw it being rewritten by default. I guess we add `sidecar.istio.io/rewriteAppHTTPProbers` just in case it is disabled in the SMCP. But yes, we do use the rewrite in the odh-dashboard.

israel-hdez commented on pkg/feature/templates/servicemesh/kserve/kserve-predictor-authorizationpolicy.tmpl:
```diff
@@ -0,0 +1,18 @@
+apiVersion: security.istio.io/v1beta1
+kind: AuthorizationPolicy
```
@retocode @skonto So, this `AuthorizationPolicy` will allow to protect `InferenceServices` behind auth (per user desire). The pod(s) of the `ISVC` will be protected, with the exception of the two paths mentioned below.

Would this be OK for KNative? I saw the queue-proxy container in the pod and was wondering if there would be any issues with it, or something else. On my naive testings, everything worked fine, but I wanted to check with you about it.

ReToCode replied:
Is this only going to run on the main containerPorts traffic (default 8080) or also on the other ones?

For example we also have traffic on:
- admin port 8022
- http2 on 8013
- metrics ports 9090/9091
- /debug/* (if you enable profiling)

Anything I've forgotten @skonto?

israel-hdez replied:
This would be intercepting everything.

So, I would infer that the default port of 8080 would be used only if the main container does not specify a port, right? From KNative upstream docs, I see that you can set a `containerPort` in the PodSpec... May that user-defined port be used transparently?? However, I see you mention 8013 for http2. Are apps required to use that for http2??? Just asking for my knowledge.

In any case, if the user can choose the port, that would be the motivation for the catch-all.

AFAIK, we can whitelist ports -- @aslakknutsen may confirm. That said, this is cluster-wide config. So, to minimize blackholes, we want to protect everything but ports/paths that would **always** work for infrastructure means.

ReToCode replied:
The ports I mentioned are on the Queue-Proxy sidecar. Knative will call the ports depending on the traffic type. So the actual user container will run as 8080 (or as defined by the user), but we'll send traffic to Queue-Proxy ports first. But if the rules are not port-related, we need to focus on the list in https://github.com/opendatahub-io/opendatahub-operator/pull/784#discussion_r1458086265

israel-hdez replied:
> but we'll send traffic to Queue-Proxy ports first.

You mean that traffic will arrive first to queue-proxy's port 8080 for HTTP1, or 8013 for HTTP2; and later is forwarded to the user container to whatever port it is defined in its spec? Asking just for my knowledge.

I just updated the `AuthorizationPolicy` (as noted in my other comment). Please, check if that would be OK for you :-)

ReToCode replied:
> You mean that traffic will arrive first to queue-proxy's port 8080 for HTTP1, or 8013 for HTTP2; and later is forwarded to the user container to whatever port it is defined in its spec? Asking just for my knowledge.

Exactly.

skonto replied:
> Anything I've forgotten @skonto?

That is fine. To add some details.
```
        // BackendHTTPPort is the backend, i.e. `targetPort` that we setup for HTTP/1 services.
	BackendHTTPPort = 8012

	// BackendHTTP2Port is the backend, i.e. `targetPort` that we setup for HTTP/2 services.
	BackendHTTP2Port = 8013

	// BackendHTTPSPort is the backend. i.e. `targetPort` that we setup for HTTPS services.
	BackendHTTPSPort = 8112

	// QueueAdminPort specifies the port number for
	// health check and lifecycle hooks for queue-proxy.
	QueueAdminPort = 8022

	// AutoscalingQueueMetricsPort specifies the port number for metrics emitted
	// by queue-proxy for autoscaler.
	AutoscalingQueueMetricsPort = 9090

	// UserQueueMetricsPort specifies the port number for metrics emitted
	// by queue-proxy for end user.
	UserQueueMetricsPort = 9091
```

BackendHTTPSPort is not being used with Istio, it is meant for internal encryption with Kourier.

For profiling if enabled:
```
	// ProfilingPort specifies the default port where profiling data is available when profiling is enabled
	ProfilingPort = 8008
```

israel-hdez commented:
> 
> > are we gonna have 605 for UI+workbench and this for kserve? for the overlapping part, should we have 605 in first then rebase for this one?
> 
> The plan was to merge this first, and rebase 605. However, we are reviewing the direction. Hold-on for now.

BTW @zdtsw, this is alive again.
The commitment is to have this PR ready and merged by 2.7 timeline. It is desired to also deliver it to 2.7, but if timing is too tight, we may delay delivery to 2.8 -- regardless of the delay, the code would still be expected to be merged by 2.7 timeline.

AFAIK, there are still no plans to merge #605 within these timelines -- although @bartoszmajsak would correct me if I'm wrong.

bartoszmajsak reviewed (changes requested):
Few smaller things, but other than that /lgtm. Sorry for round-tripping with the naming, but it just occurred to me.

I will think of what kind of tests we should write, but we can work on this together as a separate PR.

bartoszmajsak commented:
> AFAIK, there are still no plans to merge https://github.com/opendatahub-io/opendatahub-operator/pull/605 within these timelines -- although @bartoszmajsak would correct me if I'm wrong.

@israel-hdez we keep it as an organ donor and a rebase torture for now ;)

israel-hdez commented on components/kserve/kserve.go:
```diff
@@ -163,7 +165,9 @@ func (k *Kserve) ReconcileComponent(ctx context.Context, cli client.Client, resC
 	}
-	return nil
+
+	configureMeshErr := k.configureServiceMesh(dscispec)
+	return configureMeshErr
```
I don't know why not :smile:

israel-hdez commented on pkg/feature/templates/servicemesh/authorino/base/operator-cluster-wide-no-tls.tmpl:
```diff
@@ -0,0 +1,14 @@
+  name: {{ .AuthorinoConfigs.Name }}
+  namespace: {{ .Auth.Namespace }}
+spec:
+  authConfigLabelSelectors: authorino/topic=odh
```
I applied the suggestion...

Although, given both the Istio AutorizationPolicy and the Authorino AuthConfigs are quite specific to the details of KServe, I wounder if we may want to set a group other than `default`. Not sure if it may be useful some day.

bartoszmajsak replied:
They're now, but with more components coming to the picture and potential per-project authorization rules this can be helpful. I was actually thinking to prefix it with `sec.` so it will indicate `security` category. 

`sec.opendatahub.io/authorization-group: default` 

WDYT @israel-hdez @aslakknutsen

israel-hdez replied:
I'm OK with the additional prefix, but I'm not sure I like shortening it.

bartoszmajsak replied:
Good point, aligned it with other labels we use across the tools, e.g. `security.openshift.io/scc.podSecurityLabelSync`

cam-garrison commented on controllers/dscinitialization/servicemesh_setup.go:
```diff
@@ -90,18 +88,60 @@ func configureServiceMeshFeatures() feature.DefinedFeatures {
+			).
+			WithData(servicemesh.ClusterDetails).
+			PreConditions(
+				feature.EnsureCRDIsInstalled("authconfiginitializer.authorino.kuadrant.io"),
```
```suggestion
				feature.EnsureCRDIsInstalled("authconfigs.authorino.kuadrant.io"),
```

Was this intended? Assuming it was a typo/merge mistake since I don't believe authconfiginitializer  exists as a CRD.

bartoszmajsak replied:
Nopez, it's from renaming `Initializer` at some point.. in GoLand presumably as it does search text occurrences too. Guilty as charged. Thanks!

israel-hdez commented on pkg/feature/templates/servicemesh/authorino/auth-smm.tmpl:
```diff
@@ -0,0 +1,10 @@
+kind: ServiceMeshMember
+metadata:
+  name: default
+  namespace: {{ .AppNamespace }}
```
@bartoszmajsak So, with the NPs issues, I wonder if you foresee any issues because of this namespace being enrolled in the mesh?

israel-hdez replied:
Reverted to installing Authorino to its own namespace.

israel-hdez commented on config/rbac/role.yaml:
```diff
@@ -669,6 +675,15 @@ rules:
   - patch
   - update
   - watch
+- apiGroups:
+  - features.opendatahub.io
+  resources:
+  - featuretrackers/status
+  verbs:
+  - delete
+  - get
+  - patch
+  - update
```
@bartoszmajsak I'm not sure why this is missing. It is being added by `make manifests generate`.

bartoszmajsak replied:
yeah, it's been forgotten in the previous PR I believe.

israel-hdez commented on pkg/feature/templates/servicemesh/kserve/envoy-oauth-temp-fix.tmpl:
```diff
@@ -0,0 +1,76 @@
+  name: envoy-oauth-temp-fix-before
+  namespace: {{ .ControlPlane.Namespace }}
+  labels:
+    opendatahub.io/related-to: OSSM-4873
```
This ticket is already closed. I wonder if we still would need this?

bartoszmajsak replied:
@aslak can chime in here, but we have to support previous versions of OSSM?

aslakknutsen replied:
The Issue ref is why it's there, not where it will be cleaned up. I don't think we technically have a clean up task defined.

israel-hdez replied:
I guess we can leave it here, just in case somebody is using older OSSM.

BTW, I didn't understand your comment @aslakknutsen :smile:

aslakknutsen replied:
@israel-hdez The reason we have these filters is because of OSSM-4873. OSSM support a range of versions as @bartoszmajsak said, so until the cause for this has been pushed outside the support range we need it.

My previous answer was just that the issue used in the label is the cause of this filter and not the issue/task used to track the removal of it. My confusion. :)

bartoszmajsak commented on pkg/feature/servicemesh/resources.go:
```diff
@@ -0,0 +1,20 @@
+)
+
+func ConfigMaps(feature *feature.Feature) error {
+	meshConfig := feature.Spec.ControlPlane
+	if err := feature.CreateConfigMap("service-mesh-refs",
+		map[string]string{
+			"CONTROL_PLANE_NAME": meshConfig.Name,
+			"MESH_NAMESPACE":     meshConfig.Namespace,
+		}); err != nil {
+		return errors.WithStack(err)
+	}
+
+	return nil
```
@aslak I think we should populate `audiences` here too?

```suggestion
	meshConfig := feature.Spec.ControlPlane
	if err := feature.CreateConfigMap("service-mesh-refs",
		map[string]string{
			"CONTROL_PLANE_NAME": meshConfig.Name,
			"MESH_NAMESPACE":     meshConfig.Namespace,
		}); err != nil {
		return errors.WithStack(err)
	}

	authorinoConfig := feature.Spec.Auth.Authorino
	if err := feature.CreateConfigMap("auth-refs",
		map[string]string{
			"AUTH_AUDIENCE":   strings.Join(authorinoConfig.Audiences, ","),
		}); err != nil {
		return errors.WithStack(err)
	}
	
	return nil
```

aslakknutsen replied:
Oh right, this whole thing was from 605 of course..

bartoszmajsak replied:
Shh! I just rebased everything today! :D

israel-hdez replied:
Oh, well... I removed the ConfigMap in commit 540898 as I thought it wasn't needed.

aslakknutsen replied:
pushed https://github.com/opendatahub-io/opendatahub-operator/pull/784/commits/faede78d1d8669914fa556058ad804bcdc108c89

bartoszmajsak commented on pkg/feature/templates/servicemesh/authorino/auth-smm.tmpl:
```diff
@@ -0,0 +1,10 @@
+kind: ServiceMeshMember
+metadata:
+  name: default
+  namespace: {{ .AuthorinoConfigs.Namespace }}
```
I flatten it to `.Auth.Namespace` - we have it servicemesh spec already. See my overall comment for the review

bartoszmajsak commented on pkg/feature/templates/servicemesh/authorino/mesh-authz-ext-provider.patch.tmpl:
```diff
@@ -0,0 +1,13 @@
+  techPreview:
+    meshConfig:
+      extensionProviders:
+      - name: {{ .AppNamespace }}-odh-auth-provider
```
I think we can simplify names here. See my patch ^

israel-hdez replied:
So, I applied your patch in commit 773cef9.
The only thing that was a small surprise for me was updating in-place the DSCI fields. So, I though the preference was to leave the source (the DSCI or DSC) unchanged and use fields of the feature, since updating in-place may have an impact about what values other `Features` would see (and also what the controller would see once the features run).

That said, I'm not opposed on doing it. So, I simply applied the patch. Code is less confusing without the replicated fields.

bartoszmajsak reviewed (changes requested):
Thanks @israel-hdez for not only porting this code, but making it better! I think this is my final ping-pong round of the review (apologies for that!)

I made some changes locally which would be cumbersome to provide as suggestions, so I made a patch file to bundle it as a whole https://gist.github.com/bartoszmajsak/266aa0398a5cb5d0dd7f41c16a9eb4c5

israel-hdez commented on infrastructure/v1/servicemesh_types.go:
```diff
@@ -38,3 +41,22 @@ type IngressGatewaySpec struct {
+type AuthSpec struct {
+	// Namespace where it is deployed. If not provided, the default is to
+	// use '-auth-provider' suffix on the ApplicationsNamespace of the DSCI (e.g. opendatahub-auth-provider).
+	// The '-applications' suffix is removed from ApplicationsNamespace.
```
It is a little bit hard to not tell what's happening in downstream... Anyway, probably there is no reason to hide:

In downstream we have `redhat-ods-applications`, which would become `redhat-ods-auth-provider` after the manipulations which, IMO, looks better than `redhat-ods-applications-auth-provider`

israel-hdez commented on components/kserve/servicemesh_setup.go:
```diff
@@ -0,0 +1,39 @@
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/feature/servicemesh"
+)
+
+func (k *Kserve) configureServiceMesh(dscispec *dsciv1.DSCInitializationSpec) error {
```
When looking at 605, I don't even see the KServe filters.
I remember I took the filters from a specific commit somewhere. I may, simply, had forgotten to implement the cleanup.

israel-hdez replied:
I added the cleanup in 99dd030

zdtsw commented on pkg/feature/templates/servicemesh/authorino/auth-smm.tmpl:
```diff
@@ -0,0 +1,10 @@
+apiVersion: maistra.io/v1
+kind: ServiceMeshMember
+metadata:
+  name: default
```
so you have one "default" in knative-serving NS , 
another "default" in $ApplicationName-auth-provider NS

israel-hdez replied:
Yes. These are to enroll the namespaces to the Service Mesh. I think the name is not relevant.

zdtsw commented on pkg/feature/servicemesh/conditions.go:
```diff
@@ -21,6 +22,15 @@ const (
 	duration = 5 * time.Minute
 )
 
+func EnsureAuthNamespaceExists(f *feature.Feature) error {
```
so the whole reason to introduce this function instead of calling `feature.CreateNamespaceIfNotExists(...) `in PreCondition is for the `auth.namespace` value to end with `--auth-provider`?

israel-hdez replied:
I did try to use `feature.CreateNamespaceIfNotExists`. The whole history:

* When looking at other usages (like [here](https://github.com/opendatahub-io/opendatahub-operator/blob/4d8ba7ed4c6183e40ca24eb5dfa4708d853d9ea9/controllers/dscinitialization/servicemesh_setup.go#L63)) I quickly found that you need to bind the namespace value when configuring the feature. So, there is no way to use `feature.CreateNamespaceIfNotExists` directly if you need to do some processing (or maybe there is, but I didn't see it).
* To workaround the previous, I then tried using something like `WithData(ResolveAuthNamespace)` and wrapping the func, similarly to this:
```go
func (feature *Feature) error {
  feature.CreateNamespaceIfNotExists(feature.Spec.field)(feature)
}
```
and I tried to use that in `PreConditions`.
* When trying and testing the previous, I found that `PreConditions` run first and, later, `WithData` runs. This was leading to operating on an empty string because, since `WithData` runs at a later stage, the needed `feature.Spec.field` with the manipulated string is still not set.
* So, I fell back to creating this new func.

zdtsw commented on pkg/feature/servicemesh/cleanup.go:
```diff
@@ -0,0 +1,60 @@
+	for i, v := range extensionProviders {
+		extensionProvider, ok := v.(map[string]interface{})
+		if !ok {
+			log.Info("WARN: Unexpected type for extensionProvider")
```
maybe say log.Info("WARN: Unexpected type for extensionProvider will not be removed")

israel-hdez commented on controllers/dscinitialization/servicemesh_setup.go:
```diff
@@ -86,6 +86,47 @@ func configureServiceMeshFeatures() feature.FeaturesProvider {
+					//
+					// To make it part of Service Mesh we have to patch it with injection
+					// enabled instead, otherwise it will not have proxy pod injected.
+					return f.ApplyManifest(path.Join(feature.AuthDir, "deployment.injection.patch.tmpl"))
```
I think it can be debatable how to name it.
That said, I do agree that this is not really a postcondition.

I see this more like a workaround, given `Features`  currently don't have a way to do processing after applying manifests.

zdtsw replied:
sure, not saying we need to rename what we have , just food for thought, maybe we need to extend a bit in the future when we need more support in similar case

israel-hdez commented on pkg/feature/servicemesh/loaders.go:
```diff
@@ -0,0 +1,32 @@
+	dsciAuthNamespace := strings.TrimSpace(f.Spec.Auth.Namespace)
+
+	if len(dsciAuthNamespace) == 0 {
+		f.Spec.Auth.Namespace = strings.TrimSuffix(f.Spec.AppNamespace, "-applications") + "-auth-provider"
```
See the comment: https://github.com/opendatahub-io/opendatahub-operator/pull/784#discussion_r1477469889

israel-hdez commented on pkg/feature/templates/servicemesh/kserve/activator-envoyfilter.tmpl:
```diff
@@ -0,0 +1,42 @@
+kind: EnvoyFilter
+metadata:
+  labels:
+    app: odh
```
You are making me to think about it. The other `EnvoyFilters` don't  have this label.
If we want consistency, we either should remove it, or add it to the other filters.

WDYT?

zdtsw replied:
i was a bit surprised to see this label here, not really know why need it or if it is gonna be some "common" methods on others resource too. 
but in short, if you do need a label for some reason, "app:odh" is not a good one to go :D

aslakknutsen replied:
The others had it as well, but that got replaced earlier by the "clean up" labels. 

And they don't have the labels for any good reason, it was just something a threw in at as a "place holder" at some point to come up with something better later but clearly forgot. 

Maybe change it to this for now? @bartoszmajsak 

platform.opendatahub.io/component=kserve

israel-hdez replied:
So, I also don't know if we ever will need this label. However, I think we have some standard ones. I'll replace this one with the following that look standard:

```yaml
  labels:
    app.opendatahub.io/kserve: "true"
    app.kubernetes.io/part-of: kserve
```

israel-hdez commented:
@zdtsw I fixed your feedback.
Let me know if this looks ready for you.

zdtsw commented:
> @zdtsw I fixed your feedback. Let me know if this looks ready for you.

thanks! looks good to me

zdtsw commented:
@israel-hdez could you add jira ticket into PR description for tracking ?

israel-hdez commented:
> @israel-hdez could you add jira ticket into PR description for tracking ?

@zdtsw  I just added it.
I hope we can catch the upcoming ODH community release :wink:

zdtsw commented on config/crd/bases/dscinitialization.opendatahub.io_dscinitializations.yaml:
```diff
@@ -89,6 +89,28 @@ spec:
                   user experience; e.g. it provides unified authentication giving
                   a Single Sign On experience.
                 properties:
+                  auth:
```
you will need to get these change populate into `bundle` folder

zdtsw replied:
`make bundle`

zdtsw commented on components/kserve/servicemesh_setup.go:
```diff
@@ -0,0 +1,46 @@
+
+func (k *Kserve) defineServiceMeshFeatures() feature.FeaturesProvider {
+	return func(handler *feature.FeaturesHandler) error {
+		kserveExtAuthzErr := feature.CreateFeature("configure-kserve-for-external-authz").
```
similar to name as "kserve-external-authz"

zdtsw commented on components/kserve/servicemesh_setup.go:
```diff
@@ -0,0 +1,46 @@
+	return k.removeServiceMeshConfigurations(dscispec)
+}
+
+func (k *Kserve) removeServiceMeshConfigurations(dscispec *dsciv1.DSCInitializationSpec) error {
```
i did a test: to set servicemesh to Removed, but the FT "configure-kserve-for-external-authz" did not get deleted. is this correct? 
till i set kserve to Removed, FT are gone.

israel-hdez replied:
I don't think it is correct nor incorrect, but the opposite to that :smile: 

> i did a test: to set servicemesh to Removed

If there is no mesh, my expectation would be that the FT and its associated resources must be removed. I think we've been OK with dealing with the DSCI and the DSC separately. However, this is introducing some cohesion and, now, changes to the DSCI should trigger a reconcile to the DSC.

This is declared in the setup of the manager. I haven't reviewed that code.

You may do the following test: set ServiceMesh to `Removed` and, instead of setting KServe to Removed, simply add a dummy annotation to the DSC (to force its reconciliation). Most likely, you'll see that the FT will vanish.

---

Analyzing deeper, I think we shouldn't let the user to freely flip settings. E.g. setting the Mesh to `Removed` while setting Serverless to `Managed` should not be a valid state. With Vedant's PR #864 coming, the number of invalid states (and needed guards) increases.

I think we should be more proactive on helping the user and start thinking on adding a validating webhook, so that the user gets feedback on invalid configs when doing `kubectl apply`, rather than waiting for a reconciliation to happen and looking at the events/status of the DSCI/DSC. If you think this is worth to do, I can volunteer on doing the coding, since most (or all?) the things that require validation are on serving side.

bartoszmajsak replied:
> With Vedant's PR https://github.com/opendatahub-io/opendatahub-operator/pull/864 coming, the number of invalid states (and needed guards) increases.

We really need to bring validating webhook to handle such cases there instead. We could have a set of rules there and not even allow to save invalid DSC/DSCI on the cluster.

ykaliuta replied:
> > With Vedant's PR #864 coming, the number of invalid states (and needed guards) increases.
> 
> We really need to bring validating webhook to handle such cases there instead. We could have a set of rules there and not even allow to save invalid DSC/DSCI on the cluster.

webhook with one simple check has been in review for a while and you even looked in it (https://github.com/opendatahub-io/opendatahub-operator/pull/711)

israel-hdez replied:
> webhook with one simple check has been in review for a while and you even looked in it (https://github.com/opendatahub-io/opendatahub-operator/pull/711)

Ah, this is good.
So, once it is merged, I just need to add serving-specific validations.

ykaliuta replied:
> > webhook with one simple check has been in review for a while and you even looked in it (#711)
> 
> Ah, this is good. So, once it is merged, I just need to add serving-specific validations.

Join the review ;)

VaishnaviHire replied:
#711 is now merged

israel-hdez replied:
So, too late to join the review :sweat_smile:

bartoszmajsak replied:
> > > With Vedant's PR #864 coming, the number of invalid states (and needed guards) increases.
> > 
> > We really need to bring validating webhook to handle such cases there instead. We could have a set of rules there and not even allow to save invalid DSC/DSCI on the cluster.
> 
> webhook with one simple check has been in review for a while and you even looked in it (https://github.com/opendatahub-io/opendatahub-operator/pull/711)
> 

I know I did :) it's my subtle attempt to make it shipped to incubation faster. It seems it worked.

bartoszmajsak commented on pkg/feature/servicemesh/cleanup.go:
```diff
@@ -0,0 +1,66 @@
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	ctrlLog "sigs.k8s.io/controller-runtime/pkg/log"
```
```suggestion
	"sigs.k8s.io/controller-runtime/pkg/client"
```

bartoszmajsak commented on pkg/feature/servicemesh/cleanup.go:
```diff
@@ -0,0 +1,66 @@
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/gvr"
+)
+
+var log = ctrlLog.Log.WithName("features")
```
We have log in the feature itself now.

```suggestion
```

bartoszmajsak commented on pkg/feature/servicemesh/cleanup.go:
```diff
@@ -0,0 +1,66 @@
+		return err
+	}
+	if !found {
+		log.Info("no extension providers found", "feature", f.Name, "control-plane", mesh.Name, "namespace", mesh.Namespace)
```
```suggestion
		f.Log.Info("no extension providers found", "control-plane", mesh.Name, "namespace", mesh.Namespace)
```

bartoszmajsak commented on pkg/feature/servicemesh/cleanup.go:
```diff
@@ -0,0 +1,66 @@
+	for i, v := range extensionProviders {
+		extensionProvider, ok := v.(map[string]interface{})
+		if !ok {
+			log.Info("WARN: Unexpected type for extensionProvider will not be removed")
```
```suggestion
			f.Log.Info("WARN: Unexpected type for extensionProvider will not be removed")
```

bartoszmajsak commented on pkg/feature/servicemesh/cleanup.go:
```diff
@@ -0,0 +1,66 @@
+	smcp, err := f.DynamicClient.Resource(gvr.SMCP).
+		Namespace(mesh.Namespace).
+		Get(context.TODO(), mesh.Name, metav1.GetOptions{})
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			// Since the configuration of the extension provider is a patch, it could happen that
+			// the SMCP is already gone, and there will be nothing to unpatch.
+			return nil
+		}
+		return err
+	}
```
Good catch! So I will pay back for fixing my mistake with a little simplification :)

```suggestion
	if err != nil {
		// Since the configuration of the extension provider is a patch, it could happen that
		// the SMCP is already gone, and there will be nothing to unpatch.
		return client.IgnoreNotFound(err)
	}
```

bartoszmajsak commented on pkg/feature/servicemesh/cleanup.go:
```diff
@@ -0,0 +1,66 @@
+		}
+	}
+
+	_, err = f.DynamicClient.Resource(gvr.SMCP).
+		Namespace(mesh.Namespace).
+		Update(context.TODO(), smcp, metav1.UpdateOptions{})
```
For transparency, this will have to change #868, but I've got your back here.

bartoszmajsak reviewed (changes requested):
Just a few small nits :) Thanks for your patience @israel-hdez.

bartoszmajsak commented:
@zdtsw @VaishnaviHire there's one thing which is holding this PR from merging downstream. We haven't figured out yet how we want to install dependent Authorino Operator. This needs to be resolved in tandem. If we are fine with putting it in incubation and document how to use Community Operator somewhere in README here, then it's good to go.

VaishnaviHire commented:
> @zdtsw @VaishnaviHire there's one thing which is holding this PR from merging downstream. We haven't figured out yet how we want to install dependent Authorino Operator. This needs to be resolved in tandem. If we are fine with putting it in incubation and document how to use Community Operator somewhere in README here, then it's good to go.

Yes that works. Maybe we can add a README on installation of dependent operators.

israel-hdez commented on components/kserve/servicemesh_setup.go:
```diff
@@ -0,0 +1,46 @@
+		serviceMeshInitializer := feature.ComponentFeaturesHandler(k, dscispec, k.defineServiceMeshFeatures())
+		return serviceMeshInitializer.Apply()
+	}
+	if dscispec.ServiceMesh.ManagementState == operatorv1.Unmanaged && k.GetManagementState() == operatorv1.Managed {
+		return nil
+	}
```
I don't know.

My rationale was for the case when the user started with a `Managed` Mesh. Then, the user wanted to flip from `Managed` to `Unmanaged` so that odh-operator steps down and the user takes over on the setup.

Although, thinking more, probably this should also be applicable for the case when the DSCI is created with an unmanaged mesh from the beginning (i.e. they bring their own mesh). If this is the case, I think the user would be responsible for doing the setup of Mesh+Authorino+the filters involved here.

israel-hdez commented on pkg/feature/templates/servicemesh/kserve/activator-envoyfilter.tmpl:
```diff
@@ -0,0 +1,43 @@
+  name: activator-host-header
+  namespace: {{ .ControlPlane.Namespace }}
+  labels:
+    app.opendatahub.io/kserve: "true"
```
I don't know. I simply looked at KServe installation, and saw it has this label in the resources.

It seems to be injected at deployment here: https://github.com/opendatahub-io/opendatahub-operator/blob/45638fdaa235fc836ba99cc9219a14e3c7909c5a/pkg/deploy/deploy.go#L174

zdtsw replied:
looks like this template is only used for servicemesh/kserve, if it is not gonna expanded for multiple component, we could skip this label here

VaishnaviHire replied:
So that label is added using `ApplyAddLabelsPlugin` to the manifests and as bartosz pointed out is useful when a resource is a shared resource.

israel-hdez commented:
@zdtsw @bartoszmajsak I fixed your feedback.

> I'm leaning towards the view that the additional lines of code introduced by ResolveAuthNamespace function might not justify the benefits, especially so if it comes down to personal preference. So if there's no requirement nor any actual problem we are solving with this maybe it's better to just add a suffix? Less code, less reasons to worry :)

@bartoszmajsak OK, I removed the trim to the suffix.

zdtsw commented on bundle/manifests/opendatahub-operator.clusterserviceversion.yaml:
```diff
@@ -1752,7 +1777,7 @@ spec:
                 - name: DISABLE_DSC_CONFIG
                   value: "true"
-                image: REPLACE_IMAGE:latest
+                image: quay.io/edgarhz/odh-operator:latest
```
last one to go before merge:  we need revert this line

israel-hdez replied:
Ouch! I just reverted it.

israel-hdez commented:
/retest
